ENH: add weighted quantile for inverted_cdf

lorentzenchr opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
Partially solves #8935.

This PR adds support for `weights` in `np.quantile` similar to `np.average(..., weights=w)`. As an uncontroversial start, only `method="inverted_cdf"` is implemented for weights support.

See https://github.com/numpy/numpy/issues/8935#issuecomment-1382859616 for details.

lorentzenchr committed:
ENH add _weights_are_valid

lorentzenchr committed:
ENH use _weights_are_valid in np.average

lorentzenchr committed:
TST add test_quantile_constant_weights

lorentzenchr committed:
ENH add weighted quantile for inverted_cdf

lorentzenchr committed:
Merge branch 'main' into weighted_quantile_inverted_cdf

lorentzenchr committed:
CLN satisfy linter

lorentzenchr committed:
ENH add weights to nanquantile

lorentzenchr committed:
CLN make linter happy

lorentzenchr committed:
TST add test_quantile_with_integer_weights

lorentzenchr committed:
CLN remove blank line

lorentzenchr commented:
@seberg friendly ping. I know numpy 2.0 requires resources, but maybe you could have a quick look. The underlying issue is very old with a ton of discussions. (And I need help to fix the emscripten CI error.)

seberg commented on numpy/lib/function_base.py:
```diff
@@ -390,6 +390,32 @@ def iterable(y):
+        # setup wgt to broadcast along axis
+        wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
+        wgt = wgt.swapaxes(-1, axis)
+    return wgt
```
Interesting logic, but prior art in either case.

lorentzenchr replied:
Why interesting? I just shifted these lines of code out of `average` for reuse in `quantile`.

seberg replied:
The logic tries to add convenience for an N-D input with with axis and 1-D (or other input) and I find that very dubious guessing of user intent with relatively little gain (it isn't hard to make sure weights and data are properly broadcast/aligned) and potential for confusion.
(The potential for terrible bugs where you expect swapping of weights to happen but it doesn't because shapes happen to be identical seems possible but unlikely.)

lorentzenchr replied:
Again, that code is not new or from me. I just took these lines out of `np.average` for reuse in `np.quantile`.

seberg commented on numpy/lib/function_base.py:
```diff
@@ -4756,55 +4813,83 @@ def _quantile(
+        arr = np.take_along_axis(arr, index_array, axis=axis)
+        weights = np.take_along_axis(weights, index_array, axis=axis)
+        # Normalize weights to sum to one such that it corresponds to the
+        # empirical distribution function.
```
Would be tempted to throw away weights here and keep only the cummulative sum (and use  `[-1]` for the normalization.

seberg commented on numpy/lib/function_base.py:
```diff
@@ -4756,55 +4813,83 @@ def _quantile(
+            weights = np.moveaxis(weights, axis, destination=0)
+        index_array = np.argsort(arr, axis=axis, kind="stable")
+        # TODO: Which value to set to slices_having_nans.
+        slices_having_nans = None
```
Just `[-1]` after sorting.

seberg commented on numpy/lib/function_base.py:
```diff
@@ -4756,55 +4813,83 @@ def _quantile(
+        # be true.
+        # TODO: Special case quantiles == 0 and quantiles == 1?
+        arr = np.take_along_axis(arr, index_array, axis=axis)
+        weights = np.take_along_axis(weights, index_array, axis=axis)
```
This seems wrong, we should have called `moveaxis(..., axis, 0)` already, so the axis we care about is 0 and we can just use the (typically faster) `arr[index_array, ...]`.

lorentzenchr replied:
For ndim > 1, this does not seem to work anymore.

seberg commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -3765,6 +3784,43 @@ def test_quantile_add_and_multiply_constant(self, method, alpha):
+        #  - n * alpha is a float that gest rounded up
+        n = 102  # n * alpha = 20.4, 51. , 91.8
+        y = rng.random(n)
+        w = rng.integers(low=0, high=10, size=n)
```
This returns 64bit integers which repeat doesn't like apparently, I think there is `dtype=np.intp`.

seberg reviewed (commented):
I am even here unsure whether this should be called `weights` or something more clear?  For `inverted_cdf` I guess it doesn't matter (unless someone expects more than a single datapoint being mixed in the final step if its `weight < 1`, but I am not sure that can be argued to even make sense).

lorentzenchr committed:
ENH replace take_along_axis with normal indexing

lorentzenchr committed:
ENH use cumsum directly to calc CDF

lorentzenchr committed:
FIX try np.intp to fix emscripten CI

lorentzenchr committed:
FIX try np.int32 to fix CI error for np.repeat

lorentzenchr committed:
Merge branch 'main' into weighted_quantile_inverted_cdf

lorentzenchr commented:
@seberg Thank you very much for having a look!

lorentzenchr committed:
TST add test for raising errors

lorentzenchr committed:
Merge branch 'main' into weighted_quantile_inverted_cdf

lorentzenchr committed:
TST add test_quantile_with_weights_and_axis

lorentzenchr committed:
CLN make linter happy again

lorentzenchr committed:
ENH add weights to percentile

lorentzenchr commented:
@seberg I put more effort into it to also test several `ndim` of input. What should I do with `_function_base_impl.pyi`?
Can you also help with some of the failing tests? Locally with my laptop, they all pass.

lorentzenchr committed:
TYP add weights to _function_base_impl.pyi

seberg commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -3751,6 +3770,106 @@ def test_quantile_add_and_multiply_constant(self, method, alpha):
+
+    @pytest.mark.parametrize(
+            "method",
+            list(set(quantile_methods) - set(methods_supporting_weights)),
```
```suggestion
            sorted(set(quantile_methods) - set(methods_supporting_weights)),
```
I think the test failures are due to `pytest-xdist` not liking the set (random order).

Ping @BvB93 about the typing stub changes.  The seem OK to me, unless we want to restrict the method which would maybe be nice, but is there a way to do that without duplicating everything?

lorentzenchr replied:
Thanks, that fixed the CI failure.

BvB93 commented on numpy/lib/_function_base_impl.pyi:
```diff
@@ -505,6 +505,7 @@ def percentile(
     overwrite_input: bool = ...,
     method: _MethodKind = ...,
     keepdims: L[False] = ...,
+    weights: None | _ArrayLikeFloat_co = ...,
```
```suggestion
    *,
    weights: None | _ArrayLikeFloat_co = ...,
```
A minor comment: all `weights` parameters should be annotated as keyword-only. The rest looks perfectly fine though.

lorentzenchr committed:
TST sorted test parameters

lorentzenchr committed:
TYP keyword only

lorentzenchr commented:
Is there anything I can do to ease proper reviews?

seberg commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4716,55 +4821,106 @@ def _quantile(
+            i = np.searchsorted(cdf, quantiles, side="left")
+            # We might have reached the maximum with i = len(arr), e.g. for
+            # quantiles == 1.
+            i = np.minimum(i, values_count - 1)
```
This is a scalar, so I would just use an `if` or the builtin `min()`

lorentzenchr replied:
`i` now  called `indices` is an ndarray whose entries need to be `<= values_count - 1`. Therefore, `min` does not work, we need `np.minimum`.

seberg commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4716,55 +4821,106 @@ def _quantile(
+        # always, so we remove it here.
+        Nk = arr.shape[1:]
+        for kk in np.ndindex(Nk):
+            result[kk + np.s_[..., ]] = find_cdf_1d(
```
Would just use a `(...,)` rather than the fancy `s_`.  But yeah, vectorize isn't better...
(The `...` shouldn't be needed at all there?)

lorentzenchr replied:
https://github.com/numpy/numpy/pull/24254/commits/4f7841a907802cfa19d1123c526ae730ad77dafb

seberg reviewed (approved):
Well, let me just say this, I will approve the changesets (yes, probably should look a bit closer, but overall, I think its fine and adding weights to this method doesn't actually add that a lot of trickery).
Mostly also because I think the test look thorough, and I trust you on that.

The main problem remains to make the decision.  My notes on that are:
* I trust that this is correct for `inverted_cdf`.
* It is _not_ really clear to me that this is what many users really want.  (At least they can't expect something completely wrong, due to the `inverted_cdf` limitation.)
  * But: As far as I understand, you are saying that sklearn would use this.
* The definition of `weights` is non-trivial.  It is fine here, because all weights should be equivalent for the specific method.
* If someone tries to generalize `weights` to other methods, they will have to pick an "default" interpretation and/or add a name which distinguishes them.
  * (they may also have to pick an approach of how to deal with them)
* We accept that weighted quantiles are a feature that isn't ideal (should be faster) but useful enough in practice and nobody would implement a faster version locally.

**However:** I can't say I am +1 to actually merge this.  I am really only at +/-0 until there is a bit clearer "yes, this is useful in this form".

lorentzenchr commented:
> The main problem remains to make the decision.

What can I do that numpy maintainers make this decision? This original issue #8935 is over 5 years old. I can live with both yes and no, just no decision is unfortunate.

> - It is not really clear to me that this is what many users really want
>   - you are saying that sklearn would use this

Yes, scikit-learn is actually using the weighted quantile version of this PR in several places, e.g. for quantile gradient boosting, and would gladly use numpy.
See https://github.com/scikit-learn/scikit-learn/blob/e5c966e788f5b23425005100a60acbce6b0fa814/sklearn/utils/stats.py#L6

seberg commented:
Fair.  We can write a mailing list to tell everyone that we intend to merge it in this limited form and see if anyone objects.

I have been burned with quantiles before and earlier attempts at this were mostly outright wrong.  And the number of numpy devs who are really equipped to making decisions on a complicated statics problem is closer to 0 than 1.
Another thing that would help me is other devs from sklearn reviewing, e.g. since that is for me basically the same authority (maybe not when it comes to "nah, this belongs in a different package" :)).

lorentzenchr commented:
@seberg If you could take care of the mailing list, that'd be great. I ping scikit-learn devs.

@glemaitre @NicolasHug @adrinjalali @lucyleeow Friendly ping. As you have been involved in https://github.com/scikit-learn/scikit-learn/pull/17768, you might be interested in reviewing this one.

lorentzenchr commented:
@seberg Thanks for taking care of the mailing list: https://mail.python.org/archives/list/numpy-discussion@python.org/thread/JHKKUXFT6Z2KTZQZHX6EKQSLSJCFNPYM/

charris commented:
Needs a rebase.

lorentzenchr committed:
Merge branch 'main' into weighted_quantile_inverted_cdf

lorentzenchr committed:
Merge branch 'main' into weighted_quantile_inverted_cdf

lorentzenchr committed:
DOC add release note entry

seberg commented:
Anyone of the sklearn folks want to have a look?  Or are we at the point of just giving this a shot, since nobody seems to have opinions?

lorentzenchr committed:
DOC update Notes section of quantile and percentile

lorentzenchr commented:
@seberg I updated the Notes section a bit in [fd63fad](https://github.com/numpy/numpy/pull/24254/commits/fd63fad670a45784a789596a9af348077918800e). It was not 100% correct before. I hope you like it.

Concerning scikit-learn, we already have weighted quantiles, see https://github.com/scikit-learn/scikit-learn/blob/01a1fd1639be7dfe2815d42bdcb0a158bfd7365b/sklearn/utils/stats.py#L6. It is pretty much (maybe not 100%) the same as this PR.

lorentzenchr commented:
@seberg Any chance this gets included in the next release?

ogrisel commented:
> Concerning scikit-learn, we already have weighted quantiles, see https://github.com/scikit-learn/scikit-learn/blob/01a1fd1639be7dfe2815d42bdcb0a158bfd7365b/sklearn/utils/stats.py#L6. It is pretty much (maybe not 100%) the same as this PR.

I gave it a try on random data (1d inputs only) and it seems to return the same results.

ogrisel reviewed (commented):
I am running out of time for today but here is a pass of review comments. I still need to finish reading the discussions in the linked issues.

lorentzenchr commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4011,14 +4033,34 @@ def percentile(a,
+    distribution function of the data, i.e.
+    :math:`F(y) = \\frac{1}{n} \\sum_i 1_{a_i \\leq y}`.
+    Then, different methods correspond to different choices of :math:`x` that
+    fulfil the above inequalities. Methods that follow this approach are
```
fulfil is BE, fulfill AE. What's numpy's convention?

ogrisel replied:
Did not know that, feel free to ignore then.

lorentzenchr commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4011,14 +4033,34 @@ def percentile(a,
+    fulfil the above inequalities. Methods that follow this approach are
+    ``inverted_cdf`` and ``averaged_inverted_cdf``.
+
+    A more general way to define sample percentile estomators is as follows.
```
But I like tomatos :smirk:

lorentzenchr committed:
CLN address some review comments like typos

lorentzenchr committed:
Merge branch 'main' into weighted_quantile_inverted_cdf

lorentzenchr committed:
CLN fix line length

lorentzenchr committed:
FIX axis before _weights_are_valid

lorentzenchr committed:
FIX axis in weighted quantile

lorentzenchr committed:
FIX weights in _percentile_dispatcher

lorentzenchr committed:
CLN replace np.s_[..., ] by (...,)

lorentzenchr commented:
@ogrisel uncovered some shortcomings that I'm trying to solve, in particular the `out` parameter.

I'm stuck because of the following error:
```python
import numpy as np

np.take(np.arange(10), np.array([1, 4]), out=np.zeros(2, dtype=float))
```
```
TypeError: Cannot cast array data from dtype('float64') to dtype('int64') according to the rule 'safe'
```
How can I write the content of some slice of an int64 array into a float64 array?

lorentzenchr committed:
TST add weighted case to test_percentile_out

lorentzenchr committed:
CLN make out argument trivial but inefficient

lorentzenchr commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4757,55 +4881,109 @@ def _quantile(
+        # See apply_along_axis, which we do for axis=0. Note that Ni = (,)
+        # always, so we remove it here.
+        Nk = arr.shape[1:]
+        for kk in np.ndindex(Nk):
+            result[kk + (...,)] = find_cdf_1d(
+                arr[np.s_[:, ] + kk], cdf[np.s_[:, ] + kk]
+            )
```
The new tests in `test_percentile_out` fail due to this here. I'm currently unable to find a solution to make all tests pass. `result` seems to have the wrong shape in some cases.

lorentzenchr replied:
Fixed with https://github.com/numpy/numpy/pull/24254/commits/fc0b5b7a4aaefde650285e8d74885582b31749a2

lorentzenchr reviewed (commented):
~~I'm currently stuck - and I already thought the PR is ready...~~

EDIT: I finally solved it in https://github.com/numpy/numpy/pull/24254/commits/fc0b5b7a4aaefde650285e8d74885582b31749a2 and some minor fixes in https://github.com/numpy/numpy/pull/24254/commits/2d96b566401dc4d61221352a1cb03693d8ef5c7a.
I think this warrants for another proper review as those changes are not trivial at all.

lorentzenchr committed:
FIX quantile output shape

lorentzenchr committed:
FIX & TST
- cast cdf to dtype of quantile to avoid surprises
- Convert scalar python objects to pure python objects as in unweighted case
- Extend weighted case to test_linear_interpolation
- Extend weighted case to test_percentile_out

lorentzenchr committed:
CLN repacify linter

lorentzenchr committed:
ENH make out efficient again

lorentzenchr committed:
FIX nanquantile and nanpercentile

lorentzenchr commented on doc/release/upcoming_changes/24254.new_feature.rst:
```diff
@@ -0,0 +1,5 @@
+----------------------------------------------------
+The ``weights`` option is now available for `quantile`, `percentile`,
+`nanquantile` and `nanpercentile`. Only ``method="inverted_cdf"`` supports
+weights.
```
Done in https://github.com/numpy/numpy/pull/24254/commits/5581292ea141b01c6ffd31b716f4128e82caa945. Though there are no tests.

lorentzenchr replied:
Now at least some tests https://github.com/numpy/numpy/pull/24254/commits/0bbe7d9149b02df06b3215181035cbac3526655c.

lorentzenchr commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4725,55 +4840,106 @@ def _quantile(
+        r_shape = arr.shape[1:]
+        if quantiles.ndim > 0: 
+            r_shape += quantiles.shape
+        result = np.empty_like(arr, shape=r_shape)
```
First https://github.com/numpy/numpy/pull/24254/commits/5aa30c9dd662e5c8a2eea6ad2dbd08a2a1672cf2, then some important commits for fixes, then https://github.com/numpy/numpy/pull/24254/commits/5357c7786b6f9dd3fe0d435d99789246af87e24c.

lorentzenchr replied:
I might be that there is still room for improvement, e.g. by passing again the `out` arg to `take`. But I would like to postpone such optimizations to later PRs.

lorentzenchr committed:
TST add tests for nanquantile and nanpercentile

lorentzenchr committed:
FIX nan handling with slices_having_nans

lorentzenchr committed:
CLN fix BE/AE inconsistency

lorentzenchr commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4725,55 +4840,106 @@ def _quantile(
+        # TODO: Which value to set to slices_having_nans.
+        slices_having_nans = None
+        # TODO: Deal with nan values, e.g. np.isnan(arr(index_array[-1])) by
+        # be true.
```
Fixed in https://github.com/numpy/numpy/pull/24254/commits/63c3536e5de8059e12bb4c2ace667e0f9a706d8c.

I placed those TODO originally to guide reviewers to it and help me out.

lorentzenchr committed:
CLN tame the linter

lorentzenchr committed:
CLN linter again

lorentzenchr commented:
@seberg Do you think this could get in numpy 2.0? That would be my hope.

seberg commented:
Thanks, @lorentzenchr and @ogrisel for reviewing.  The new changes look good, and the new tests were really needed.  I did have a look at the shape logic, and it seems really quite clean to me, only comments I added were so small, that there would be no point in following up.

Since voiced anything about the added API here or on the mailing list.  Putting this in.
